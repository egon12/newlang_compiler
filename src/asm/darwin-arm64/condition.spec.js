const { Generator } = require('./literal_asm.js');
const fs = require('fs');

it('should generate condition codes', () => {
	const g = new Generator();
	const result = g.generate(node);
	expect(result).toMatchSnapshot();
	fs.writeFileSync(__dirname + '/condition.spec.s', result);
})

const node = [
	{
		type: 'FunctionDefinition',
		namespace: 'main',
		name: 'main',
		parameters: [],
		body: [
			{
				type: 'VariableDefinition',
				name: 'a',
				value: { type: 'Number', value: "4" },
			},
			{
				type: 'IfStatement',
				condition: {
					type: 'MathBoolExpression',
					operator: '==',
					left: { type: 'Number', value: "4" },
					right: { type: 'Variable', id: "a" },
				},
				body: [
					{
						type: 'ReturnStatement',
						value: { type: 'Number', value: 4, }
					}
				],
			},
			{
				type: 'ReturnStatement',
				value: { type: 'Number', value: 0, }
			}
		]
	}
]


