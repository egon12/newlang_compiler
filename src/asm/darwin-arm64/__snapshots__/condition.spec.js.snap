// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate condition codes 1`] = `
".global _main
.align 2

_main:
	bl main.main
	mov x16, #1
	svc 0x80
main.main:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, x29, #16
	mov x0, #4
	str x0, [x29, #-8]
	mov x0, #4
	ldr x1, [x29, #-8]
	cmp x0, x1
	beq main.main.if1.true
	b main.main.if1.else
main.main.if1.true:
	mov x0, #4
	b main.main.return
main.main.if1.else:
	mov x0, #0
main.main.return:
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret
data:


"
`;
